### DISCUSSION TITLE
Best Practice to auto generate navigation based on /pages

### DISCUSSION THREAD

**Discussion by seilex** [nuxt/nuxt#16640] [Questions]
I'm pretty new to nuxt3 and currently taking my first steps. I had the idea to create an boilerplate to kick start future projects to gain some experience. I could solve most of my challenges but stumbled upon the creation of an proper navigation component. 

**This is where I am:**

1. I use the "/pages" directory to create the page structure. 
```
/sectionA
--childA1.vue
--childA2.vue
/sectionB
--childB1.vue
/sectionC
--childC1.vue
sectionA.vue
sectionB.vue
sectionC.vue
index.vue
```

2. I use "/layouts" and have an "public" layout which is used for non-authenticated users. As soon as you're logged in the default layout with my navigation components should be used. The base structure of the layout should be basically something like this:
```
// default.vue
<template>
  <div>
    <NavBar v-if="mainNav" :data="mainNav"/>
    <h1>{{sectionTitle}}</h1>
    <NavTabs v-if="sectionChildren" :data="sectionChildren"/>
    <slot/>
  </div>
</template>
```

3. I use nested routes so my app.vue looks like this 

```
// app.vue
<template>
  <div>
    <NuxtLayout>
      <NuxtPage />
    </NuxtLayout>
  </div>
</template>
```

4. Both navigation components are built with Tailwind UI and require an array with the navigation items
```
const mainNav = [
      { name: "Index", href: "/", current: false },
      { name: "SectionA", href: "/sectionA", current: true },
      { name: "SectionB", href: "/sectionB", current: false },
      { name: "SectionC", href: "/sectionC", current: false },
    ];

const indexPageChildren = [
      { name: "childA1", href: "/sectionA/childA1", current: true },
      { name: "childA2", href: "/sectionA/childA2", current: false },
    ];
```

**This is what I want to achieve:**
1. I want to create the required array with the navigation items from the nuxt routes. Whenever the user navigates within the page the array needs to be updated, so that the navigation component can show the right active states. 
2. Each section has an default child page. So if the user for examples browses to "/sectionA" I would like to forward him to the default child page (``/sectionA/childB`` for example)

I've tried a few approaches but no one really led to success - I was able to generate the array from useRouter but couldn't find any solution that ensures that the navigation components are updated properly when navigating through the site. Since I believe this is not something total unusual I'm a bit scared that I'm on the complete wrong path. So I would appreciate any help/hints pointing me in the right direction. Of course I'm also happy about feedback regarding the general approach. 

Thanks

**Comment by andrewdleach** [comment]
Not sure if it's best practice, but this is the pattern I've been using to generate navigation based on pages in the `/content` directory:

```
<script setup>
const { data: navigation } = await useAsyncData('navigation', () => fetchContentNavigation())
</script>
```

And then use `navigation` as a prop in a sub component (that will also get mounted recursively for sub trees in the nav)

```
<template>
    <ul tabindex="0" :class=presentation>
        <li v-for="item in links">
            <NuxtLink :to="item._path">
                {{ item.title }}
            </NuxtLink>
            <NavigationComponent v-if="item.children" :navigation-tree="item.children" />
        </li>
    </ul>
</template>
```

Source: [Fetch Content Navigation](https://content.nuxtjs.org/api/composables/fetch-content-navigation/)
